### PR TITLE
Remove hardwired roles inclusion

### PR DESCRIPTION
**Pull request by ccamacho**
This commit allows to dinamically include
all the roles in the repository so the Ansible
interpreter can find them independently of
how many roles are in the repository.

### CODE REVIEW

File: config/ansible.cfg [hunk 1]
```diff
@@ -13,8 +13,11 @@ retry_files_enabled = False
 retry_files_save_path = /tmp/ansible-installer-retries
 nocows = True
 remote_user = root
-# cd roles; echo ../roles/* | tr " " :
-roles_path = ../roles/benchmarking:../roles/cluster:../roles/codeflare:../roles/entitlement:../roles/gpu_operator:../roles/load_aware:../roles/local_ci:../roles/nfd:../roles/ocm:../roles/pipelines:../roles/rhods:../roles/utils:../roles/wisdom:../roles/watsonx_serving:../roles/llm_load_test:../roles/notebooks
+
+# Both roles_path and collections_paths are configured
+# dinamically in the_common.py library, this is to avoid
```
Comment (kpouget) [inline config/ansible.cfg:18]:
```suggestion
# dynamically in the_common.py library, this is to avoid
```

File: topsail/_common.py [hunk 2]
```diff
@@ -162,6 +162,21 @@ def _run(self):
         print(f"Using '{env['ANSIBLE_CACHE_PLUGIN_CONNECTION']}' to store ansible facts.")
         Path(env["ANSIBLE_CACHE_PLUGIN_CONNECTION"]).parent.mkdir(parents=True, exist_ok=True)
 
+        # We configure the roles path dinamycally appending them to the defaults
+        top_dir_roles = os.path.join(top_dir, 'roles')
+        top_dir_roles_list = [os.path.join(top_dir_roles, entry) for entry in os.listdir(top_dir_roles) if os.path.isdir(os.path.join(top_dir_roles, entry))]
```
Comment (kpouget) [inline topsail/_common.py:167]: Can you use `pathlib`?
```
top_dir_roles = top_dir / "roles"
```
and maybe:
```
top_dir_roles_list = top_dir_roles.glob("*")
```
(can be filtered with`os.path.isdir`)

File: topsail/_common.py [hunk 3]
```diff
@@ -162,6 +162,21 @@ def _run(self):
...
+        for path in top_dir_roles_list:
+            env["ANSIBLE_ROLES_PATH"] = os.path.join(env.get("ANSIBLE_ROLES_PATH", ""), path)
+        self.ansible_vars["roles_path"] = env["ANSIBLE_ROLES_PATH"]
+
+        # We configure the collections path dinamycally
```
Comment (kpouget) [inline topsail/_common.py:172]:
```suggestion
        # We configure the collections path dynamically
```

File: topsail/_common.py (hunk 2 above)
Comment (ccamacho) [reply topsail/_common.py:167]: sure thing pathlib is modern stuff

File: topsail/_common.py [hunk 4]
```diff
@@ -162,6 +162,23 @@ def _run(self):
         print(f"Using '{env['ANSIBLE_CACHE_PLUGIN_CONNECTION']}' to store ansible facts.")
         Path(env["ANSIBLE_CACHE_PLUGIN_CONNECTION"]).parent.mkdir(parents=True, exist_ok=True)
 
+        # We configure the roles path dynamically appending them to the defaults
+        top_dir_roles = Path(top_dir / "roles")
```
Comment (kpouget) [inline topsail/_common.py:166]:
```
top_dir = Path(__file__).resolve().parent.parent
```
`top_dir is already a Path, so:
```suggestion
        top_dir_roles = top_dir / "roles"
```

File: topsail/_common.py [hunk 5]
```diff
@@ -162,6 +162,23 @@ def _run(self):
...
+        # top_dir_roles_list = [os.path.join(top_dir_roles, entry) for entry in os.listdir(top_dir_roles) if os.path.isdir(os.path.join(top_dir_roles, entry))]
```
Comment (kpouget) [inline topsail/_common.py:167]: can be removed

File: topsail/_common.py [hunk 6]
```diff
@@ -162,6 +162,23 @@ def _run(self):
...
+        top_dir_roles_list = [str(entry) for entry in top_dir_roles.glob('*') if entry.is_dir()]
+        current_roles_path = env.get("ANSIBLE_ROLES_PATH", "")
+        env["ANSIBLE_ROLES_PATH"] = os.pathsep.join([current_roles_path] + top_dir_roles_list)
+        self.ansible_vars["roles_path"] = env["ANSIBLE_ROLES_PATH"]
+
+        # We configure the collections path dynamically
+        current_collections_paths = [env.get("ANSIBLE_COLLECTIONS_PATHS", "tikitiki")]
```
Comment (kpouget) [inline topsail/_common.py:174]:
```suggestion
        current_collections_paths = []
        if (collect_path := env.get("ANSIBLE_COLLECTIONS_PATHS")); collect_path is not None:
          current_collections_paths.append(collect_path)
```

File: topsail/_common.py [hunk 7]
```diff
@@ -162,6 +162,23 @@ def _run(self):
...
+        for path in sys.path:
+            collections_path = str(Path(path) / 'ansible_collections')
+            if os.path.exists(collections_path):
+                current_collections_paths = current_collections_paths + [collections_path]
```
Comment (kpouget) [inline topsail/_common.py:178]:
```suggestion
            collections_path = Path(path) / 'ansible_collections'
            if collections_path.exists():
                current_collections_paths.append(collections_path)
```

Comment (kpouget) [inline topsail/_common.py:178]:
```
current_collections_paths = current_collections_paths + [collections_path]
```
this form ^^^ should be avoided,  this one vvv is prefered:
```
current_collections_paths += [collections_path]
```
and this ^^^ and this vvv is equivalent:
```
current_collections_paths.append(collections_path)
```
I use one or the other depending on the wind, not sure if there's any strong argument to prefer one side over the other

**Comment by kpouget** [comment]
the notebook-light [failure](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_topsail/110/pull-ci-openshift-psap-topsail-main-notebooks-light/1726578926832062464/artifacts/light/test/artifacts/003__plots/report_00_report:_error_report.html) is a known flake, the test passed ✅ 
/override ci/prow/notebooks-light

**Comment by kpouget** [comment]
thanks for the PR & rewritings @ccamacho, merging the PR :)

**Pull request merged**